src/lang-promql/lsp/client.ts

```diff
@@ -0,0 +1,61 @@
+import axios from 'axios';
+import { AutocompleteContext, Completion, CompletionResult } from "@codemirror/next/autocomplete";
+import { AutocompleteResponse, TextEdit } from "./model";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function isIterable(obj: any): boolean {
+  return obj != null && typeof obj[ Symbol.iterator ] === 'function';
+}
+
+export class LSPClient {
+  private readonly url: string
+  private readonly autocompleteEndpoint = "/completion"
+
+  constructor(url: string) {
+    this.url = url;
+  }
+
+  autocomplete(context: AutocompleteContext): Promise<CompletionResult> {
+    const {state, pos} = context
+    const line = state.doc.lineAt(pos)
+    const body = {
+      expr: state.doc.sliceString(0),
+      limit: 100,
+      // positionChar start at 0 at the beginning of the line for LSP
+      positionChar: pos - line.from - 1 > 0 ? pos - line.from - 1 : pos - line.from,
```

Comment:
Are you sure these are counted in UTF-16 Codepoints?

If not, the positions will be off in case the query includes non-ASCII characters.

Reply:
no idea actually, But if it's not the case, I believe the bug would be more on CodeMirror side

Reply:
Not necessarily.

There is an incredible number of ways to encode positions, and UTF16 Codepoints is one of the weirder ones. Unfortunately that's what the LSP uses.

It could well be (I haven't looked it up), that Codemirror uses a different way of counting chars.

Reply:
I just had a quick look at Codemirror's code, and while it doesn't appear to be documented anywhere, they indeed use UTF16 code points.

Reply:
nice, thanks !